In [3]:
import pandas as pd
import external_func as ef
import scipy.stats as st

In [4]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_demo.txt"
df_final_demo = pd.read_csv(url, sep=",")
df_final_demo.head(3)

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
0,836976,6.0,73.0,60.5,U,2.0,45105.30,6.0,9.0
1,2304905,7.0,94.0,58.0,U,2.0,110860.30,6.0,9.0
2,1439522,5.0,64.0,32.0,U,2.0,52467.79,6.0,9.0


In [5]:
url_2 = "https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_experiment_clients.txt"
df_final_experiment_clients = pd.read_csv(url_2, sep=",")
df_final_experiment_clients.head(3)

,client_id,Variation
0,9988021,Test
1,8320017,Test
2,4033851,Control


In [6]:
url_pt_1 = "https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_1.txt"
df_pt_1 = pd.read_csv(url_pt_1)
df_pt_1.head(3)

,client_id,visitor_id,visit_id,process_step,date_time
0,9988021,580560515_7732621733,781255054_21935453173_531117,step_3,2017-04-17 15:27:07
1,9988021,580560515_7732621733,781255054_21935453173_531117,step_2,2017-04-17 15:26:51
2,9988021,580560515_7732621733,781255054_21935453173_531117,step_3,2017-04-17 15:19:22


In [7]:
url_pt_2 = "https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_2.txt"
df_pt_2 = pd.read_csv(url_pt_2)
df_pt_2.head(3)

,client_id,visitor_id,visit_id,process_step,date_time
0,763412,601952081_10457207388,397475557_40440946728_419634,confirm,2017-06-06 08:56:00
1,6019349,442094451_91531546617,154620534_35331068705_522317,confirm,2017-06-01 11:59:27
2,6019349,442094451_91531546617,154620534_35331068705_522317,step_3,2017-06-01 11:58:48


## Analysis of df_final_experiment_clients dataframe

In [8]:
df_final_experiment_clients.shape

(70609, 2)

In [9]:
df_final_experiment_clients.dtypes

client_id    int64
Variation      str
dtype: object

In [10]:
df_final_experiment_clients.isna().sum()

client_id        0
Variation    20109
dtype: int64

In [11]:
df_final_experiment_clients = ef.columns_formating(df_final_experiment_clients)
df_final_experiment_clients

,client_id,variation
0,9988021,Test
1,8320017,Test
2,4033851,Control
3,1982004,Test
4,9294070,Control
...,...,...
70604,2443347,NaN
70605,8788427,NaN
70606,266828,NaN
70607,1266421,NaN


In [12]:
df_variation_nan = df_final_experiment_clients[df_final_experiment_clients["variation"].isna()]
df_variation_nan

,client_id,variation
50500,5459747,NaN
50501,8031000,NaN
50502,1847030,NaN
50503,9713157,NaN
50504,7775828,NaN
...,...,...
70604,2443347,NaN
70605,8788427,NaN
70606,266828,NaN
70607,1266421,NaN


In [18]:
df_final_experiment_clients["variation"].value_counts()

variation
Test       26968
Control    23532
Name: count, dtype: int64

In [13]:
df_final_experiment_clients["variation"].unique()

<StringArray>
['Test', 'Control', nan]
Length: 3, dtype: str

**Conclusions:**

- Data types are ok for this dataset
- Besides NaN We have two values of "variation" variable: 'Test', 'Control'. We don't need to clear the data in this column
- NaNs: 20109 -> 28,5% of dataset -> we're not aware of the participation of almost 30% of clients
- Test: 26968 -> 38,2% of dataset -> approximately 38% of clients included in test group
- Control: 23532 -> 33,32% of dataset -> approximately 33% of clients included in control group

## Analysis of df_final_demo dataframe

In [21]:
df_final_demo.shape

(70609, 9)

In [22]:
df_final_demo.dtypes

client_id             int64
clnt_tenure_yr      float64
clnt_tenure_mnth    float64
clnt_age            float64
gendr                   str
num_accts           float64
bal                 float64
calls_6_mnth        float64
logons_6_mnth       float64
dtype: object

In [23]:
df_final_demo["gendr"].unique()

<StringArray>
['U', 'M', 'F', nan, 'X']
Length: 5, dtype: str

**Conclusion.**
 - M: Male / Man
 - F: Female / Woman
 - X: Non-binary, intersex, or gender-diverse
 - U: Unspecified, unknown, or unrecorded

In [24]:
df_final_demo.isna().sum()

client_id            0
clnt_tenure_yr      14
clnt_tenure_mnth    14
clnt_age            15
gendr               14
num_accts           14
bal                 14
calls_6_mnth        14
logons_6_mnth       14
dtype: int64

In [25]:
df_gender_nan = df_final_demo[df_final_demo["gendr"].isna()]
df_gender_nan

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
4164,7402828,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8316,355337,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8677,8412164,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
13444,2222915,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18066,4876926,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25961,5277910,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
28432,7616759,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
35323,8191345,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
43518,1227228,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
46076,8611797,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [33]:
clients_without_group = df_final_demo[df_final_demo["client_id"].isin(df_variation_nan["client_id"])]
clients_without_group

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
50500,414870,6.0,80.0,57.0,U,2.0,492304.33,5.0,5.0
50501,1047811,9.0,118.0,30.0,M,2.0,27085.88,5.0,5.0
50502,6090170,5.0,62.0,28.0,U,2.0,39855.30,4.0,4.0
50503,706631,16.0,193.0,56.0,U,2.0,52298.64,5.0,5.0
50504,6844077,16.0,196.0,48.5,F,2.0,518763.48,7.0,7.0
...,...,...,...,...,...,...,...,...,...
70604,7993686,4.0,56.0,38.5,U,3.0,1411062.68,5.0,5.0
70605,8981690,12.0,148.0,31.0,M,2.0,101867.07,6.0,6.0
70606,333913,16.0,198.0,61.5,F,2.0,40745.00,3.0,3.0
70607,1573142,21.0,255.0,68.0,M,3.0,475114.69,4.0,4.0


We can identify clients (by client_id) with missing participation information

In [22]:
c_df.isna().sum()

client_id           0
clnt_tenure_yr      2
clnt_tenure_mnth    2
clnt_age            2
gendr               2
num_accts           2
bal                 2
calls_6_mnth        2
logons_6_mnth       2
dtype: int64

For two of them we have no information about their activity in the app

**Conclusions:**
- We have 14 clients with empty data (by all rows)
- Data types are ok for this dataset
- Important to notice - we have 4 different options in "gender" category
- Clients who participated in the experiment 'df_final_experiment_clients' without specifying a group can be found in the dataset 'df_final_demo'

**Open questions:**
- Can we identify criterias for spliting groups? -> Can identify the group for clients with NaNs from 'df_final_experiment_clients' -> Do we really need to identify the group for NaN
- What client activity from 'df_final_demo' is more valuabl for us?

### What can we say about the clients who participated/not participated in the test?

In [27]:
test_group = df_final_experiment_clients[df_final_experiment_clients["variation"] == "Test"]["client_id"]
test_group

0        9988021
1        8320017
3        1982004
5        9466803
6        3170143
          ...   
50490    5969816
50492    8692087
50497    7230446
50498    5230357
50499    6334360
Name: client_id, Length: 26968, dtype: int64

In [28]:
control_group = df_final_experiment_clients[df_final_experiment_clients["variation"] == "Control"]["client_id"]
control_group

2        4033851
4        9294070
7        6651403
9        2105948
12       9814849
          ...   
50491    4364429
50493    8730282
50494    5305116
50495     393005
50496    2908510
Name: client_id, Length: 23532, dtype: int64

In [29]:
not_participated_clients = df_final_demo[~df_final_demo["client_id"].isin(df_final_experiment_clients["client_id"])]
not_participated_clients

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth


**Conclusion 1** 

We can say, that all clients from df_final_demo are participated in test. We need to identify the groups for NaN client_id's

In [30]:
clients_in_test_gr = df_final_demo[df_final_demo["client_id"].isin(test_group)]
clients_in_test_gr

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
0,836976,6.0,73.0,60.5,U,2.0,45105.30,6.0,9.0
2,1439522,5.0,64.0,32.0,U,2.0,52467.79,6.0,9.0
3,1562045,16.0,198.0,49.0,M,2.0,67454.65,3.0,6.0
7,388801,30.0,361.0,57.5,M,5.0,522498.72,1.0,4.0
9,8198645,15.0,189.0,54.5,F,2.0,382303.83,6.0,9.0
...,...,...,...,...,...,...,...,...,...
50492,501496,23.0,281.0,35.0,F,2.0,32529.34,0.0,3.0
50495,1780858,21.0,262.0,68.5,M,3.0,372100.59,6.0,9.0
50497,5826160,20.0,249.0,56.5,F,2.0,44837.16,2.0,5.0
50498,8739285,19.0,229.0,69.5,F,2.0,44994.24,1.0,4.0


In [31]:
clients_in_control_gr = df_final_demo[df_final_demo["client_id"].isin(control_group)]
clients_in_control_gr

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
1,2304905,7.0,94.0,58.0,U,2.0,110860.30,6.0,9.0
4,5126305,12.0,145.0,33.0,F,2.0,103671.75,0.0,3.0
5,3727881,5.0,71.0,30.5,U,2.0,23915.60,0.0,3.0
6,272934,5.0,66.0,58.5,U,2.0,27021.42,2.0,5.0
8,285619,30.0,369.0,67.5,M,2.0,299388.72,3.0,6.0
...,...,...,...,...,...,...,...,...,...
50486,6662941,7.0,84.0,36.0,F,2.0,83101.84,4.0,7.0
50491,2787512,23.0,283.0,48.0,F,3.0,123618.60,1.0,4.0
50493,9932367,23.0,280.0,57.0,F,2.0,94466.55,6.0,9.0
50494,4079649,23.0,276.0,62.5,M,3.0,57575.28,4.0,7.0


In [34]:
clients_without_group[["clnt_tenure_yr", "clnt_tenure_mnth", "clnt_age", "num_accts", "bal", "calls_6_mnth", "logons_6_mnth"]].describe()

,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,num_accts,bal,calls_6_mnth,logons_6_mnth
count,20107.000000,20107.000000,20107.000000,20107.000000,2.010700e+04,20107.000000,20107.000000
mean,12.106232,151.271746,44.238773,2.257920,1.422490e+05,4.108619,4.147710
std,6.900585,82.451745,15.555909,0.538317,3.001242e+05,2.192047,2.178329
min,2.000000,33.000000,13.500000,1.000000,1.378942e+04,1.000000,1.000000
25%,6.000000,82.000000,30.500000,2.000000,3.014772e+04,2.000000,2.000000
50%,11.000000,138.000000,45.000000,2.000000,5.633810e+04,4.000000,4.000000
75%,16.000000,193.000000,56.500000,2.000000,1.305127e+05,6.000000,6.000000
max,62.000000,749.000000,92.500000,8.000000,1.283842e+07,7.000000,7.000000


In [35]:
clients_in_test_gr[["clnt_tenure_yr", "clnt_tenure_mnth", "clnt_age", "num_accts", "bal", "calls_6_mnth", "logons_6_mnth"]].describe()

,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,num_accts,bal,calls_6_mnth,logons_6_mnth
count,26961.000000,26961.000000,26961.000000,26961.000000,2.696100e+04,26961.000000,26961.000000
mean,11.982901,149.852899,47.163922,2.249917,1.489626e+05,3.061941,6.101851
std,6.844947,81.783390,15.514461,0.529335,3.073737e+05,2.190590,2.177119
min,2.000000,33.000000,17.000000,1.000000,2.378944e+04,0.000000,3.000000
25%,6.000000,82.000000,33.500000,2.000000,3.972920e+04,1.000000,4.000000
50%,11.000000,134.000000,47.500000,2.000000,6.546836e+04,3.000000,6.000000
75%,15.000000,191.000000,59.500000,2.000000,1.394786e+05,5.000000,8.000000
max,55.000000,669.000000,96.000000,7.000000,1.632004e+07,6.000000,9.000000


In [36]:
clients_in_control_gr[["clnt_tenure_yr", "clnt_tenure_mnth", "clnt_age", "num_accts", "bal", "calls_6_mnth", "logons_6_mnth"]].describe()

,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,num_accts,bal,calls_6_mnth,logons_6_mnth
count,23527.000000,23527.000000,23526.000000,23527.000000,2.352700e+04,23527.000000,23527.000000
mean,12.087686,151.060186,47.498427,2.259914,1.501473e+05,3.129213,6.166277
std,6.877531,82.126443,15.521781,0.538559,2.958070e+05,2.184502,2.173014
min,2.000000,33.000000,17.000000,2.000000,2.378961e+04,0.000000,3.000000
25%,6.000000,82.000000,33.500000,2.000000,4.012335e+04,1.000000,4.000000
50%,11.000000,137.000000,48.500000,2.000000,6.602418e+04,3.000000,6.000000
75%,16.000000,192.000000,60.000000,2.000000,1.406060e+05,5.000000,8.000000
max,55.000000,669.000000,96.000000,7.000000,8.292996e+06,6.000000,9.000000


In [41]:
display(clients_without_group.isna().sum())
display(clients_in_test_gr.isna().sum())
display(clients_in_control_gr.isna().sum())

client_id           0
clnt_tenure_yr      2
clnt_tenure_mnth    2
clnt_age            2
gendr               2
num_accts           2
bal                 2
calls_6_mnth        2
logons_6_mnth       2
dtype: int64

client_id           0
clnt_tenure_yr      7
clnt_tenure_mnth    7
clnt_age            7
gendr               7
num_accts           7
bal                 7
calls_6_mnth        7
logons_6_mnth       7
dtype: int64

client_id           0
clnt_tenure_yr      5
clnt_tenure_mnth    5
clnt_age            6
gendr               5
num_accts           5
bal                 5
calls_6_mnth        5
logons_6_mnth       5
dtype: int64

**Conclusion 2**

We have three gpoups of clients: in "test" group, in "control" group and without group_name. We can see almost the same statistical data for each group

**Conclusion 3**
We have a missing values for 14 clients. We need to clarify, if we need to drop them or they have a valuable data

## Analysis for df_final_web_data

In [53]:
df_pt_1.shape

(343141, 5)

In [79]:
df_pt_1.isna().sum()

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

In [91]:
df_pt_1.duplicated().sum()

np.int64(2095)

In [54]:
df_pt_2.shape

(412264, 5)

In [80]:
df_pt_2.isna().sum()

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

In [92]:
df_pt_2.duplicated().sum()

np.int64(8669)

In [84]:
merged_df = pd.merge(df_pt_2, df_pt_1, on=["client_id", "visitor_id", "visit_id", "process_step", "date_time"], how='outer')
merged_df

,client_id,visitor_id,visit_id,process_step,date_time
0,169,201385055_71273495308,749567106_99161211863_557568,confirm,2017-04-12 20:23:09
1,169,201385055_71273495308,749567106_99161211863_557568,start,2017-04-12 20:19:36
2,169,201385055_71273495308,749567106_99161211863_557568,step_1,2017-04-12 20:19:45
3,169,201385055_71273495308,749567106_99161211863_557568,step_2,2017-04-12 20:20:31
4,169,201385055_71273495308,749567106_99161211863_557568,step_3,2017-04-12 20:22:05
...,...,...,...,...,...
755400,9999875,738878760_1556639849,931268933_219402947_599432,start,2017-06-01 22:40:01
755401,9999875,738878760_1556639849,931268933_219402947_599432,step_1,2017-06-01 22:40:08
755402,9999875,738878760_1556639849,931268933_219402947_599432,step_1,2017-06-01 22:41:28
755403,9999875,738878760_1556639849,931268933_219402947_599432,step_2,2017-06-01 22:41:47


In [85]:
print(merged_df["client_id"].isna().sum())
print(merged_df["visitor_id"].isna().sum())
print(merged_df["visit_id"].isna().sum())
print(merged_df["process_step"].isna().sum())
print(merged_df["date_time"].isna().sum())

0
0
0
0
0


In [93]:
merged_df.duplicated().sum()

np.int64(10764)

In [86]:
343141+412264

755405

**Conclusion:**
- After merging we've got dataframe with data from df_pt_1 and df_pt_2 
- The amount of potential duplicates is a bit more than a sum of duplicates from each dataset

In [101]:
nan_group_clients = merged_df[merged_df["client_id"].isin(df_variation_nan["client_id"])]
nan_group_clients.shape

(128522, 5)

Here we have information about client activity with NaN group name and after filtering out duplicated rows:

In [124]:
nan_group_clients[nan_group_clients.duplicated(subset=['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']) == False]

,client_id,visitor_id,visit_id,process_step,date_time
249,2078,585735301_21309149782,900598259_99565669243_552843,start,2017-05-06 08:29:00
251,2078,585735301_21309149782,900598259_99565669243_552843,start,2017-05-06 08:36:18
253,2078,585735301_21309149782,900598259_99565669243_552843,start,2017-05-06 08:36:40
255,2078,585735301_21309149782,900598259_99565669243_552843,start,2017-05-06 08:37:14
257,2078,585735301_21309149782,900598259_99565669243_552843,step_1,2017-05-06 08:29:28
258,2078,585735301_21309149782,900598259_99565669243_552843,step_1,2017-05-06 08:36:24
259,2078,585735301_21309149782,900598259_99565669243_552843,step_1,2017-05-06 08:36:46
260,2078,585735301_21309149782,900598259_99565669243_552843,step_1,2017-05-06 08:37:19
261,2078,585735301_21309149782,900598259_99565669243_552843,step_2,2017-05-06 08:30:08
262,2078,585735301_21309149782,900598259_99565669243_552843,step_3,2017-05-06 08:31:39


In [110]:
nan_group = nan_group_clients[nan_group_clients.duplicated(subset=['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']) == False]
nan_group.shape

(126662, 5)

In [95]:
test_group_clients = merged_df[merged_df["client_id"].isin(test_group)]
test_group_clients.shape

(177847, 5)

Here we have information about client activity with 'Test' group name and after filtering out duplicated rows:

In [125]:
test_group = test_group_clients[test_group_clients.duplicated(subset=['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']) == False]
test_group.shape

(176699, 5)

In [96]:
control_group_clients = merged_df[merged_df["client_id"].isin(test_group)]
control_group_clients.shape

(177847, 5)

Here we have information about client activity with 'Control' group name and after filtering out duplicated rows:

In [126]:
control_group = control_group_clients[control_group_clients.duplicated(subset=['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']) == False]
control_group.shape

(176699, 5)

**Conclusion**
- We have activity for all 14 clients. But at this step we're not sure if we have valuable data for all of them -> we can't drop clients witn NaN client info

In [135]:
test_group.head(3)

,client_id,visitor_id,visit_id,process_step,date_time
12,555,402506806_56087378777,637149525_38041617439_716659,confirm,2017-04-15 13:00:34
13,555,402506806_56087378777,637149525_38041617439_716659,start,2017-04-15 12:57:56
14,555,402506806_56087378777,637149525_38041617439_716659,step_1,2017-04-15 12:58:03


In [173]:
test_group["t_combo_id"] = (test_group["client_id"].astype(str)+"-"+test_group["visitor_id"].astype(str)+"-"+test_group["visit_id"].astype(str))
test_group

,client_id,visitor_id,visit_id,process_step,date_time,combo_id,t_combo_id
12,555,402506806_56087378777,637149525_38041617439_716659,confirm,2017-04-15 13:00:34,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
13,555,402506806_56087378777,637149525_38041617439_716659,start,2017-04-15 12:57:56,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
14,555,402506806_56087378777,637149525_38041617439_716659,step_1,2017-04-15 12:58:03,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
15,555,402506806_56087378777,637149525_38041617439_716659,step_2,2017-04-15 12:58:35,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
16,555,402506806_56087378777,637149525_38041617439_716659,step_3,2017-04-15 13:00:14,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
...,...,...,...,...,...,...,...
755376,9999729,843385170_36953471821,493310979_9209676464_421146,start,2017-04-20 14:28:57,9999729-843385170_36953471821-493310979_920967...,9999729-843385170_36953471821-493310979_920967...
755377,9999729,843385170_36953471821,493310979_9209676464_421146,step_1,2017-04-20 14:22:49,9999729-843385170_36953471821-493310979_920967...,9999729-843385170_36953471821-493310979_920967...
755378,9999729,843385170_36953471821,493310979_9209676464_421146,step_2,2017-04-20 14:27:36,9999729-843385170_36953471821-493310979_920967...,9999729-843385170_36953471821-493310979_920967...
755391,9999832,145538019_54444341400,472154369_16714624241_585315,start,2017-05-16 16:46:03,9999832-145538019_54444341400-472154369_167146...,9999832-145538019_54444341400-472154369_167146...


In [175]:
control_group["contr_combo_id"] = (control_group["client_id"].astype(str)+"-"+control_group["visitor_id"].astype(str)+"-"+control_group["visit_id"].astype(str))
control_group

,client_id,visitor_id,visit_id,process_step,date_time,combo_id,contr_combo_id
12,555,402506806_56087378777,637149525_38041617439_716659,confirm,2017-04-15 13:00:34,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
13,555,402506806_56087378777,637149525_38041617439_716659,start,2017-04-15 12:57:56,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
14,555,402506806_56087378777,637149525_38041617439_716659,step_1,2017-04-15 12:58:03,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
15,555,402506806_56087378777,637149525_38041617439_716659,step_2,2017-04-15 12:58:35,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
16,555,402506806_56087378777,637149525_38041617439_716659,step_3,2017-04-15 13:00:14,555-402506806_56087378777-637149525_3804161743...,555-402506806_56087378777-637149525_3804161743...
...,...,...,...,...,...,...,...
755376,9999729,843385170_36953471821,493310979_9209676464_421146,start,2017-04-20 14:28:57,9999729-843385170_36953471821-493310979_920967...,9999729-843385170_36953471821-493310979_920967...
755377,9999729,843385170_36953471821,493310979_9209676464_421146,step_1,2017-04-20 14:22:49,9999729-843385170_36953471821-493310979_920967...,9999729-843385170_36953471821-493310979_920967...
755378,9999729,843385170_36953471821,493310979_9209676464_421146,step_2,2017-04-20 14:27:36,9999729-843385170_36953471821-493310979_920967...,9999729-843385170_36953471821-493310979_920967...
755391,9999832,145538019_54444341400,472154369_16714624241_585315,start,2017-05-16 16:46:03,9999832-145538019_54444341400-472154369_167146...,9999832-145538019_54444341400-472154369_167146...


In [178]:
grouped_activities = test_group.groupby(['process_step']).agg(test_cnt_client=('t_combo_id', 'count'))
grouped_activities

,test_cnt_client
process_step,
confirm,25600
start,55773
step_1,38666
step_2,30899
step_3,25761


In [180]:
control_groupped_activities = control_group.groupby(['process_step']).agg(control_cnt_client=('contr_combo_id', 'count'))
control_groupped_activities

,control_cnt_client
process_step,
confirm,25600
start,55773
step_1,38666
step_2,30899
step_3,25761


In [ ]:
# KPI's
# The time online
# the amount of click's


In [99]:
#Here we can find activity of 14 clients, who has no data in 'df_final_demo'
merged_df_copy = merged_df[merged_df["client_id"].isin(df_gender_nan["client_id"])]
merged_df_copy["client_id"].value_counts()

client_id
5144725    14
1227228    13
1037867    12
7402828    12
8611797    10
4876926     8
8191345     8
1618749     7
7616759     7
6021001     6
8412164     6
355337      5
5277910     5
2222915     2
Name: count, dtype: int64